libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,210 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+
+using namespace std;
+using namespace solidity::frontend;
+
+namespace {
+
+bool nodeSet(FunctionCallGraphBuilder::Node const& _node)
+{
+	if (auto node = std::get_if<FunctionCallGraphBuilder::SpecialNode>(&_node))
+		return *node != FunctionCallGraphBuilder::SpecialNode::Unset;
+
+	return true;
+}
+
+}
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	m_graph = make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::CreationRoot;
+	m_currentDispatch = SpecialNode::CreationDispatch;
+	visitConstructor(
+		_contract,
+		_contract.annotation().linearizedBaseContracts.cbegin() + 1,
+		_contract.annotation().linearizedBaseContracts.cend()
+	);
+	m_currentNode = SpecialNode::Unset;
+	m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
```

Comment:
Is this really there for *all* publicly reachable functions? What about getters for public state variables? I'm not saying we necessarily need to cover implicitly generated functions like getters here, but if we choose to ignore them, maybe at least mention it in a comment.

Reply:
Looking at the runtime dispatch loop below: can't you just call ``add`` here as well? Why do it differently?

Reply:
> Looking at the runtime dispatch loop below: can't you just call add here as well? Why do it differently?

There is a huge difference between `visitCallable` and `add`. `add` simply adds the node according to the given parameters. `visitCallable` on the other hand actually visits all calls that the given function makes..

The runtime dispatch loop is just synchronizing the two dispatches to have the same nodes, no further visits required. the interface function loop on the other hand is where the actual work happens, where it visits all the functions and the functions that those call in turn..

> Is this really there for all publicly reachable functions? What about getters for public state variables? 

I don't think there is a need to visit the getters as they don't call any other functions..

Reply:
> I don't think there is a need to visit the getters as they don't call any other functions..

But you still want to verify that there is code generated for them in codegen, don't you? Also they are called *by* other functions...

Reply:
When they are called _by_, they will be recorded as functions that were called... (leaf nodes)
But it is note-worthy that as the code currently is, verification succeeds.

Reply:
Verification succeeds, because you don't verify getters :-). Which *are* part of the external interface, so not sure why they shouldn't be verified as well...

Reply:
Granted, though, it's not a major issue, but for completeness I'd include them both in the graph and in the verification.

Reply:
If you *want* to verify them in codegen note that their names begin with ``"getter_fun_"`` not with ``"fun_"`` - that's why they're silently ignored so far.

Reply:
Marking this resolved, since apparently we don't care too much about getters.

Reply:
Please document it in the header.